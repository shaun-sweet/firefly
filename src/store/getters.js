export default {
  user (state) {
    return {
      displayName: state.displayName,
      email: state.email,
      id: state.id
    }
  },

  homesList (state) {
    return Object.keys(state.homes).map(homeId => {
      return {
        label: state.homes[homeId].nickname,
        value: homeId
      }
    })
  },

  selectedHome (state) {
    return state.selectedHome
  },

  isLoadingInitialState (state) {
    return state.appState.isLoadingInitialState
  },

  defaultHome (state) {
    return Object.keys(state.homes).filter(home => state.homes[home].role === 'owner')[0]
  },

  devicesViewList (state, getters) {
    if (!getters.isLoadingInitialState) {
      const devicesViewList = state.homes[state.selectedHome].devicesViewList
      const devicesArray = Object.keys(devicesViewList).map((deviceId) => {
        const device = devicesViewList[deviceId]
        // this is what auto complete fuzzy finder search feature needs to display the properties
        device.stamp = device.primaryStateStatus
        device.label = device.alias
        return device
      })
      return devicesArray.filter((device) => device.export_ui)
    }
    return []
  }

  // notifications (state, getters) {
  //   const selectedHome = state.homes[getters.selectedHome]
  //   return Object.keys(selectedHome.notifications).map(id => selectedHome.notifications[id])
  // },

  // events (state, getters) {
  //   const selectedHome = state.homes[getters.selectedHome]
  //   return Object.keys(selectedHome.events).map(id => selectedHome.events[id])
  // }
}
