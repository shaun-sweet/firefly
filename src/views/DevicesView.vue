<template>
  <div class="layout-padding">
    <q-modal 
      class="device-modal"
      :content-css="{minWidth: '60vw', padding: '1em'}" 
      @close="closeDeviceMenuHandler()" 
      ref="deviceMenu"
    >
      <!-- <div class="layout-padding"> -->
        <action
          v-for="(action, index) in deviceModalActions"
          :key="index"
          :actionMetadata="action"
         />
        <q-btn class="close-btn" color="primary" @click="$refs.deviceMenu.close()">Close</q-btn>
      <!-- </div> -->
    </q-modal>
    <div class="row search-bar">
      <div class="col-6">
        <q-search v-model="searchTerms" placeholder="Search a device"> 
          <q-autocomplete
            separator
            :filter="fuzzyFind"
            :static-data="{field: 'alias', list: devicesViewList}"
            @search="fuzzyFind"
            @selected="selected"
          />
        </q-search>
      </div>
    </div>
    <div class="row">
      <device 
        v-for="device in devicesViewList"
        :key="device.ff_id"
        :title="device.alias"
        :device-id="device.ff_id"
        :on-click-handler="onDeviceMenuHandler"
        :device-metadata="device.metadata"
      />
    </div>
  </div>
</template>

<script>
import Fuse from 'fuse.js'
import Device from '@/Device'
import Action from '@/Action'

export default {
  components: {
    Device,
    Action
  },
  computed: {
    devicesViewList () {
      return this.$store.getters.devicesViewList.sort((a, b) => {
        if (a.alias < b.alias) return -1
        if (a.alias > b.alias) return 1
        return 0
      })
    },
    deviceModalActions () {
      return this.$store.getters.deviceModalActions
    }
  },
  methods: {
    onDeviceMenuHandler (deviceId, deviceMetadata) {
      const payload = {
        deviceId,
        deviceMetadata
      }
      this.$store.dispatch('openDeviceMenu', payload)
      this.$refs.deviceMenu.open()
    },
    closeDeviceMenuHandler () {
      this.$store.dispatch('closeDeviceMenu')
    },
    fuzzyFind () {
      const fuzzyFinder = new Fuse(this.devicesViewList, this.fuzzyFindoptions)
      return fuzzyFinder.search(this.searchTerms)
    },
    selected (item) {
      return item
    }
  },
  data () {
    return {
      searchTerms: '',
      checked: false,
      fuzzyFindoptions: {
        shouldSort: true,
        threshold: 0.6,
        location: 0,
        distance: 100,
        maxPatternLength: 32,
        minMatchCharLength: 1,
        keys: [
          "alias"
        ]
      }
    }
  }
}
</script>

<style lang="stylus" scoped>
.options-menu 
  margin-left 0.2em
  cursor pointer

.close-btn
  margin-left 8px

</style>
