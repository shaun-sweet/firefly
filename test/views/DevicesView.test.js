import { shallow, createLocalVue } from 'vue-test-utils'
import Vuex from 'vuex'
import DevicesView from 'src/views/DevicesView.vue'
import getters from 'src/store/getters'
import {initialState, testState} from '../fixtures/state'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('DevicesView', () => {
  let wrapper;
  let store;
  let actions;
  let state;
  let spy;
  beforeEach(() => {
    spy = jest.fn()
    state = {...testState}
    actions = {
      closeDeviceMenu: jest.fn(),
      openDeviceMenu: jest.fn(() => new Promise((resolve, reject) => resolve()))
    }
    store = new Vuex.Store({
      state,
      actions,
      getters
    })
    wrapper = shallow(DevicesView,{ store, localVue });
  })
  it('should render the DevicesView correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should call closeDeviceMenuHandler when modal is closed', () => {
    store.hotUpdate({
      actions: { closeDeviceMenu: spy }
    })
    wrapper.find('q-modal').trigger('close');
    expect(spy).toHaveBeenCalled()
  });

  xit('should call closeDeviceMenuHandler when button targeting deviceMenu ref', () => {
    store.hotUpdate({
      actions: { closeDeviceMenu: spy }
    })
    wrapper.find({ref: 'deviceMenu' }).trigger('close');

    // $refs does't work with the vue-test-utils rendered component
    // This is basically testing the samething as above assuming $refs triggers the dom element correctly

    expect(spy).toHaveBeenCalled()
  });

  it('should update deviceSettings.alias on input', () => {
    wrapper.vm.onSaveHandler = spy;
    const text = 'test alias'
    const input = wrapper.find('q-input')
    input.element.value = text
    input.trigger('change')
    wrapper.setData({
      deviceSettings: {
        alias: text
      }
    });
    expect(wrapper.vm.deviceSettings.alias).toBe(text)

    const saveBtn = wrapper.findAll('q-btn').at(1)
    saveBtn.trigger('click')

    expect(spy).toHaveBeenCalled()
  });


})
