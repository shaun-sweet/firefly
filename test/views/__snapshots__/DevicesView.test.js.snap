// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevicesView should render the DevicesView correctly 1`] = `
<div
  class="layout-padding"
>
  <q-modal
    class="device-modal"
    content-css="[object Object]"
  >
    <q-tabs
      class="tab-container"
      no-pane-border=""
    >
      <q-tab
        default="default"
        label="Controls"
        name="controls"
      />
       
      <q-tab
        label="Settings"
        name="settings"
      />
       
      <q-tab
        label="About"
        name="about"
      />
       
      <q-icon
        class="x-close-button desktop-only"
        name="ion-close"
      />
       
      <q-tab-pane
        name="controls"
      >
        <q-list
          class="controls"
          no-border=""
        >
          <q-list-header>
            <h5>
              Controls
            </h5>
          </q-list-header>
           
          <div
            class="controls-container"
          >
            <!---->
            <!---->
             
            <q-btn
              class="close-btn desktop-only"
              color="primary"
            >
              Close
            </q-btn>
          </div>
        </q-list>
      </q-tab-pane>
       
      <q-tab-pane
        name="settings"
      >
        <q-list
          class="settings"
          no-border=""
        >
          <q-list-header>
            <h5>
              Settings
            </h5>
          </q-list-header>
        </q-list>
         
        <form>
          <q-field>
            <q-input
              stack-label="Device Alias"
              type="text"
            />
          </q-field>
           
          <q-field>
            <q-btn
              color="primary"
              icon="ion-ios-checkmark-outline"
            >
              Save
            </q-btn>
             
            <q-btn
              icon="ion-ios-arrow-thin-left"
              outline=""
            >
              Cancel
            </q-btn>
             
            <q-btn
              class="delete-device-btn float-right gt-md"
              color="red"
              icon="ion-ios-trash"
            >
              Delete
            </q-btn>
          </q-field>
        </form>
      </q-tab-pane>
       
      <q-tab-pane
        name="about"
      >
        <q-list
          class="about"
          no-border=""
        >
          <q-list-header>
            <h5>
              About
            </h5>
          </q-list-header>
           
          <p>
            Device Alias: test alias
          </p>
           
          <p>
            Device UUID: xyz123
          </p>
           
          <p>
            Device Info:
          </p>
           
          <q-list
            no-border=""
            separator=""
          >
            <q-item
              class="device-view__item"
              tag="p"
            >
              <strong>
                STATUS
              </strong>
              : ok
              
            </q-item>
          </q-list>
        </q-list>
      </q-tab-pane>
    </q-tabs>
     
    <q-btn
      class="fixed mobile-only"
      color="primary"
      round=""
      style="left: 18px; bottom: 18px;"
    >
      <q-icon
        name="ion-arrow-left-c"
      />
    </q-btn>
  </q-modal>
   
  <div
    class="row search-bar"
  >
    <div
      class="col-6"
    >
      <q-search
        placeholder="Search a device"
      >
        <q-autocomplete
          filter="function boundFn(a) {
    var l = arguments.length;
    return l
      ? l > 1
        ? fn.apply(ctx, arguments)
        : fn.call(ctx, a)
      : fn.call(ctx)
  }"
          separator=""
          static-data="[object Object]"
        />
      </q-search>
    </div>
  </div>
   
  <div
    class="row"
  >
    <!---->
    <!---->
  </div>
</div>
`;
